[BEAM-4747] mkdirs if they don't exist in localfilesystem

ryan-williams opened the pull request:
Change `LocalFileSystem` to match semantics of e.g. `GCSFileSystem`: writing to a path in a non-existent directory should just create the intermediate directories, instead of throwing `IOError: [Errno 2] No such file or directory`

R: @lukecwik 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

lukecwik commented on sdks/python/apache_beam/io/localfilesystem.py:
```diff
@@ -127,6 +127,9 @@ def _path_open(self, path, mode, mime_type='application/octet-stream',
     """Helper functions to open a file in the provided mode.
     """
     compression_type = FileSystem._get_compression_type(path, compression_type)
+    parent = os.path.dirname(path)
```
We should only create the path in the `create` call and not the `open` call as we'll get a weird error if the user mistypes the path for something being read and we will try to create the directory which may fail (e.g. permissions) which will raise a confusing error message.

Do you want to add a test to localfilesystem_test.py so that this isn't regressed?

ryan-williams commented:
Good point, probably worth fixing in `rename`/`copy` as well.

Confusingly, I'm now seeing the portable wordcount example not fail without this change… let me try to get an answer one way or another there, and likely file a different JIRA to link this against, as well as addressing your changes.

If I confirm that this isn't actually an issue in the portable wordcount example, and that means you don't think it's worth making this change at all, let me know, thanks!

ryan-williams commented:
OK, I'm reorienting this PR around [BEAM-4747](https://issues.apache.org/jira/browse/BEAM-4747) and will respond to your comments above shortly.

lukecwik commented:
@ryan-williams Any updates?

ryan-williams commented:
Sorry I dropped this, looking into it now!

ryan-williams commented:
OK, I moved the dir-creation to apply on `create` but not `open`, added similar logic to `rename` and `copy`, and tried to add tests for all of them in a way that made sense. lmk what you think!

charlesccychen commented:
This PR breaks the 2.7.0 RC2 because it doesn't take into account relative paths.  I will therefore be reverting this.

elwinarens commented on sdks/python/apache_beam/io/localfilesystem.py:
```diff
@@ -217,6 +224,10 @@ def rename(self, source_file_names, destination_file_names):
+        parent = os.path.dirname(destination)
+        if not os.path.exists(parent):
+          os.makedirs(parent)
+
```
Could have used `os.renames` instead
